// @blockfrost/blockfrost-js/lib/endpoints/api/epochs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.epochsParameters = exports.epochsBlocksByPoolId = exports.epochsBlocks = exports.epochsStakesByPoolId = exports.epochsStakes = exports.epochsPrevious = exports.epochsNext = exports.epochsLatest = exports.epochs = void 0;
const utils_1 = require("../../../utils");
async function epochs(number) {
    return new Promise((resolve, reject) => {
        this.axiosInstance(`${this.apiUrl}/epochs/${number}`)
            .then(resp => {
            resolve(resp.data);
        })
            .catch(err => {
            reject(utils_1.handleError(err));
        });
    });
}
exports.epochs = epochs;
async function epochsLatest() {
    return new Promise((resolve, reject) => {
        this.axiosInstance(`${this.apiUrl}/epochs/latest`)
            .then(resp => {
            resolve(resp.data);
        })
            .catch(err => {
            reject(utils_1.handleError(err));
        });
    });
}
exports.epochsLatest = epochsLatest;
async function epochsNext(number, pagination) {
    const paginationOptions = utils_1.getPaginationOptions(pagination);
    return new Promise((resolve, reject) => {
        this.axiosInstance(`${this.apiUrl}/epochs/${number}/next`, {
            params: {
                page: paginationOptions.page,
                count: paginationOptions.count,
                order: paginationOptions.order,
            },
        })
            .then(resp => {
            resolve(resp.data);
        })
            .catch(err => {
            reject(utils_1.handleError(err));
        });
    });
}
exports.epochsNext = epochsNext;
async function epochsPrevious(number, pagination) {
    const paginationOptions = utils_1.getPaginationOptions(pagination);
    return new Promise((resolve, reject) => {
        this.axiosInstance(`${this.apiUrl}/epochs/${number}/previous`, {
            params: {
                page: paginationOptions.page,
                count: paginationOptions.count,
                order: paginationOptions.order,
            },
        })
            .then(resp => {
            resolve(resp.data);
        })
            .catch(err => {
            reject(utils_1.handleError(err));
        });
    });
}
exports.epochsPrevious = epochsPrevious;
async function epochsStakes(number, pagination) {
    const paginationOptions = utils_1.getPaginationOptions(pagination);
    return new Promise((resolve, reject) => {
        this.axiosInstance(`${this.apiUrl}/epochs/${number}/stakes`, {
            params: {
                page: paginationOptions.page,
                count: paginationOptions.count,
                order: paginationOptions.order,
            },
        })
            .then(resp => {
            resolve(resp.data);
        })
            .catch(err => {
            reject(utils_1.handleError(err));
        });
    });
}
exports.epochsStakes = epochsStakes;
async function epochsStakesByPoolId(number, poolId, pagination) {
    const paginationOptions = utils_1.getPaginationOptions(pagination);
    return new Promise((resolve, reject) => {
        this.axiosInstance(`${this.apiUrl}/epochs/${number}/stakes/${poolId}`, {
            params: {
                page: paginationOptions.page,
                count: paginationOptions.count,
                order: paginationOptions.order,
            },
        })
            .then(resp => {
            resolve(resp.data);
        })
            .catch(err => {
            reject(utils_1.handleError(err));
        });
    });
}
exports.epochsStakesByPoolId = epochsStakesByPoolId;
async function epochsBlocks(number, pagination) {
    const paginationOptions = utils_1.getPaginationOptions(pagination);
    return new Promise((resolve, reject) => {
        this.axiosInstance(`${this.apiUrl}/epochs/${number}/blocks`, {
            params: {
                page: paginationOptions.page,
                count: paginationOptions.count,
                order: paginationOptions.order,
            },
        })
            .then(resp => {
            resolve(resp.data);
        })
            .catch(err => {
            reject(utils_1.handleError(err));
        });
    });
}
exports.epochsBlocks = epochsBlocks;
async function epochsBlocksByPoolId(number, poolId, pagination) {
    const paginationOptions = utils_1.getPaginationOptions(pagination);
    return new Promise((resolve, reject) => {
        this.axiosInstance(`${this.apiUrl}/epochs/${number}/blocks/${poolId}`, {
            params: {
                page: paginationOptions.page,
                count: paginationOptions.count,
                order: paginationOptions.order,
            },
        })
            .then(resp => {
            resolve(resp.data);
        })
            .catch(err => {
            reject(utils_1.handleError(err));
        });
    });
}
exports.epochsBlocksByPoolId = epochsBlocksByPoolId;
async function epochsParameters(number) {
    return new Promise((resolve, reject) => {
        this.axiosInstance(`${this.apiUrl}/epochs/${number}/parameters`)
            .then(resp => {
            resolve(resp.data);
        })
            .catch(err => {
            reject(utils_1.handleError(err));
        });
    });
}
exports.epochsParameters = epochsParameters;
