// @blockfrost/blockfrost-js/lib/endpoints/api/metadata/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.metadataTxsLabelCbor = exports.metadataTxsLabel = exports.metadataTxsLabels = void 0;
const utils_1 = require("../../../utils");
async function metadataTxsLabels(pagination) {
    const paginationOptions = utils_1.getPaginationOptions(pagination);
    return new Promise((resolve, reject) => {
        this.axiosInstance(`${this.apiUrl}/metadata/txs/labels`, {
            params: {
                page: paginationOptions.page,
                count: paginationOptions.count,
                order: paginationOptions.order,
            },
        })
            .then(resp => {
            resolve(resp.data);
        })
            .catch(err => {
            reject(utils_1.handleError(err));
        });
    });
}
exports.metadataTxsLabels = metadataTxsLabels;
async function metadataTxsLabel(label, pagination) {
    const paginationOptions = utils_1.getPaginationOptions(pagination);
    return new Promise((resolve, reject) => {
        this.axiosInstance(`${this.apiUrl}/metadata/txs/labels/${label}`, {
            params: {
                page: paginationOptions.page,
                count: paginationOptions.count,
                order: paginationOptions.order,
            },
        })
            .then(resp => {
            resolve(resp.data);
        })
            .catch(err => {
            reject(utils_1.handleError(err));
        });
    });
}
exports.metadataTxsLabel = metadataTxsLabel;
async function metadataTxsLabelCbor(label, pagination) {
    const paginationOptions = utils_1.getPaginationOptions(pagination);
    return new Promise((resolve, reject) => {
        this.axiosInstance(`${this.apiUrl}/metadata/txs/labels/${label}/cbor`, {
            params: {
                page: paginationOptions.page,
                count: paginationOptions.count,
                order: paginationOptions.order,
            },
        })
            .then(resp => {
            resolve(resp.data);
        })
            .catch(err => {
            reject(utils_1.handleError(err));
        });
    });
}
exports.metadataTxsLabelCbor = metadataTxsLabelCbor;
